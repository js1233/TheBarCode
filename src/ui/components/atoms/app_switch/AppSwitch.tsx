import React, { useEffect, useState } from "react";
import {
  // StyleProp,
  StyleSheet,
  SwitchProps
  // ViewStyle
} from "react-native";
import { AppLog, TAG } from "utils/Util";
import ToggleSwitch from "toggle-switch-react-native";
import { usePreferredTheme } from "hooks";
import SwitchActive from "assets/images/switch_active.svg";
import SwitchInActive from "assets/images/switch_inactive.svg";

export interface AppButtonProps extends SwitchProps {
  defaultValue: boolean;
  onValueChange: (isSwitchEnabled: boolean) => void;
  // style?: StyleProp<ViewStyle>;
  showCustomThumb?: boolean;
  shouldNeedCallBackCreate?: boolean;
  shouldToggle?: boolean;
}

export const AppSwitch = React.memo<AppButtonProps>(
  ({
    defaultValue,
    onValueChange,
    showCustomThumb = false,
    shouldNeedCallBackCreate = false,
    shouldToggle = true,
    // style,
    ...rest
  }) => {
    const [isEnabled, setIsEnabled] = useState<boolean>(defaultValue);
    const { themedColors } = usePreferredTheme();

    useEffect(() => {
      if (!shouldNeedCallBackCreate) {
        onValueChange(defaultValue);
      }
    }, [shouldNeedCallBackCreate, defaultValue, onValueChange]);

    useEffect(() => {
      if (shouldToggle) {
        setIsEnabled(defaultValue);
      }
    }, [shouldToggle, defaultValue]);

    const toggleSwitch = () => {
      if (!shouldToggle) {
        AppLog.log(() => "AppSwitch() => toggle working", TAG.VENUE);
        onValueChange(!isEnabled);
      } else {
        setIsEnabled((previousState) => !previousState);
        onValueChange(!isEnabled);
      }
    };
    // houldToggleTrue === false ? isEnabled : !isEnabled
    return (
      <ToggleSwitch
        testID="app-switch"
        isOn={isEnabled}
        onColor={themedColors.primaryColor}
        offColor={themedColors.secondaryColor}
        onToggle={toggleSwitch}
        thumbOffStyle={styles.thumb}
        icon={
          showCustomThumb ? (
            isEnabled ? (
              <SwitchActive width={20} height={20} />
            ) : (
              <SwitchInActive width={20} height={20} />
            )
          ) : null
        }
        {...rest}
      />
    );
  }
);

const styles = StyleSheet.create({
  switch: {},
  thumb: {},
  tinyLogo: {
    width: 20,
    height: 20
  }
});
