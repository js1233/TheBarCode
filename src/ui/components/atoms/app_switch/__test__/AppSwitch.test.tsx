import renderer from "react-test-renderer";
import { AppSwitch } from "ui/components/atoms/app_switch/AppSwitch";
import React from "react";
import { fireEvent, render } from "@testing-library/react-native";

it("Renders snapshot as expected", () => {
  // given
  const onValueChangeCallback = jest.fn();
  const tree = renderer
    .create(
      <AppSwitch
        defaultValue={true}
        onValueChange={onValueChangeCallback}
      />
    )
    .toJSON();

  // then
  expect(tree).toMatchSnapshot();
});

describe("test if callbacks works as expected", () => {
  it("check switch on when initial value of switch is false", () => {
    // given
    const onValueChangeCallback = jest.fn();
    const { getByTestId } = render(
      <AppSwitch
        defaultValue={false}
        onValueChange={onValueChangeCallback}
      />
    );

    // when
    const appSwitch = getByTestId("app-switch");
    fireEvent(appSwitch, "onToggle");

    // then
    expect(onValueChangeCallback).toHaveBeenCalledTimes(2);
    expect(onValueChangeCallback).toHaveBeenLastCalledWith(true);
  });

  it("works fine with given initial value", () => {
    // given
    let currentValue: boolean | undefined;
    const onValueChangeCallback = jest.fn((_value: boolean) => {
      currentValue = _value;
    });

    // when
    const { getByTestId } = render(
      <AppSwitch
        defaultValue={true}
        onValueChange={onValueChangeCallback}
      />
    );
    const appSwitch = getByTestId("app-switch");

    // then
    expect(currentValue).toBe(true);

    // when
    fireEvent(appSwitch, "onToggle");

    // then
    expect(currentValue).toBe(false);
  });
});
