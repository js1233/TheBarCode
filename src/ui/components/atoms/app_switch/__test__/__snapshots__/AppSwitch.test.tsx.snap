// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders snapshot as expected 1`] = `
<
  icon={null}
  isOn={true}
  offColor="#D4D4D4"
  onColor="#0F766E"
  onToggle={[Function]}
  testID="app-switch"
  thumbOffStyle={Object {}}
/>
`;
